### PR TITLE
add 1559 support

### PR DESCRIPTION
**Pull request by dorothy-zbornak**
## Description

- Add 1559 fields (`maxFeePerGas`, `maxPriorityFeePerGas`) to transaction and call types.
- Add fee market transaction support to `PrivateKeyWalletSupbrovider` and `MnemonicWalletSubprovider`.
  - On networks that support 1559 (ropsten, rinkeby, mainnet), 1559 will be preferred if either `maxFeePerGas` or `maxPriorityFeePerGas` exists. 
  - On networks that support 2930 (ropsten, rinkeby, kovan, mainnet), 2930 will be preferred if `accessList` exists.
- Add new rpc methods to `Web3Wrapper`: `getGasPriceAsync()`, `getBaseFeePerGasAsync()`, and `getMaxPriorityFeePerGasAsync()` (not supported by metamask).
  - `getMaxPriorityFeePerGasAsync()` won't work on metamask so the more universal way to get that value is to do `getGasPriceAsync() - getBaseFeePerGasAsync()`.  HOWEVER, there is no guarantee that `getGasPriceAsync() >= getBaseFeePerGasAsync() + getMaxPriorityFeePerGasAsync()` so you can wind up with a negative max priority fee, so clamp it or something :trollface:.
- Updated `MetamaskSupbrovider` to:
  - auto-detect network and strip out conflicting gas price fields (MM will get angry if both `gasPrice` and 1559 fields are passed in).
  - auto-forward RPC calls to the underlying MM provider (why wasn't it doing this before?) 

Some findings:
- MM will always send out 1559 txs on eligible networks, regardless of what you pass in (it will try to deduce missing fields), but it will throw up if you pass in `gasPrice` + 1559 fields, which is why the `MetamaskSubprovider` will now strip them out.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: subproviders/src/subproviders/metamask_subprovider.ts [hunk 1]
```diff
@@ -131,4 +162,34 @@ export class MetamaskSubprovider extends Subprovider {
...
+    private async _getCommonAsync(): Promise<Common> {
+        if (this._common) {
+            return this._common;
+        }
+        const chainId = await this._web3Wrapper.getChainIdAsync();
+        return (this._common = getCommonForChain(chainId));
+    }
+}
+
+function cleanTxDataForChain<TRPCData extends TxDataRPC | CallDataRPC, TData extends TxData | CallData>(
```
Comment (dorothy-zbornak) [inline subproviders/src/subproviders/metamask_subprovider.ts:175]: MM will throw up if you pass in a mix of legacy and 1559 fee fields.

File: web3-wrapper/src/web3_wrapper.ts [hunk 2]
```diff
@@ -219,6 +219,33 @@ export class Web3Wrapper {
...
+            params: ['pending', false],
+        });
+        const { baseFeePerGas } = rawBlock;
+        return new BigNumber(baseFeePerGas || 0);
+    }
+    /**
+     * Fetch the current max piority fee per gas. This is the suggested miner tip
+     * to get mined in the current block.
+     */
+    public async getMaxPriorityFeePerGasAsync(): Promise<BigNumber> {
```
Comment (dorothy-zbornak) [inline web3-wrapper/src/web3_wrapper.ts:245]: This does not work on MM :disappointed:

File: web3-wrapper/src/web3_wrapper.ts [hunk 3]
```diff
@@ -219,6 +219,33 @@ export class Web3Wrapper {
         const chainId = _.parseInt(chainIdStr);
         return chainId;
     }
+    /**
+     * Fetch the current gas price.
+     * For post-London hardfork chains, this will be baseFeePerGas + maxPriorityFeePerGas
```
Comment (dorothy-zbornak) [inline web3-wrapper/src/web3_wrapper.ts:224]: *usually

File: web3-wrapper/src/web3_wrapper.ts [hunk 4]
```diff
@@ -219,6 +219,33 @@ export class Web3Wrapper {
...
+    /**
+     * Fetch the base fee per gas for the pending block.
+     */
+    public async getBaseFeePerGasAsync(): Promise<BigNumber> {
+        const rawBlock = await this.sendRawPayloadAsync<{ baseFeePerGas?: string }>({
+            method: 'eth_getBlockByNumber',
+            params: ['pending', false],
+        });
+        const { baseFeePerGas } = rawBlock;
+        return new BigNumber(baseFeePerGas || 0);
```
Comment (dorothy-zbornak) [inline web3-wrapper/src/web3_wrapper.ts:239]: Field might not exist on all networks.

File: monorepo-scripts/src/test_installation.ts [hunk 5]
```diff
@@ -132,13 +132,14 @@ async function testInstallPackageAsync(
...
             module: 'commonjs',
-            target: 'es5',
+            target: 'es6',
             lib: ['es2017', 'dom'],
             declaration: true,
             noImplicitReturns: true,
             pretty: true,
             strict: true,
             resolveJsonModule: true,
+            esModuleInterop: true,
```
Comment (dorothy-zbornak) [inline monorepo-scripts/src/test_installation.ts:142]: `ethereumjs-util` now uses ES-style imports in its dist files.

File: subproviders/src/subproviders/metamask_subprovider.ts [hunk 6]
```diff
@@ -56,16 +60,32 @@ export class MetamaskSubprovider extends Subprovider {
                     end(err);
                 }
                 return;
-            case 'eth_sendTransaction':
-                const [txParams] = payload.params;
+            case 'eth_call': {
```
Comment (dekz) [inline subproviders/src/subproviders/metamask_subprovider.ts:63]: This was from legacy where you would only use Metamask for the signing component and provide your own RPC provider to do everything else.

Back then during ICO days Metamask's connection would be down 90% of the time 😂

**Review by dekz — APPROVED** [review]

File: subproviders/src/subproviders/private_key_wallet.ts [hunk 7]
```diff
@@ -133,3 +119,31 @@ export class PrivateKeyWalletSubprovider extends BaseWalletSubprovider {
...
 }
+
+/**
+ * Create a transaction object based on the default network Common and transaction params.
+ */
+function createTransactionObject(
+    common: Common,
+    txParams: PartialTxParams,
+): ethjsTx.Transaction | ethjsTx.FeeMarketEIP1559Transaction | ethjsTx.AccessListEIP2930Transaction {
+    const normalziedTxParams = { ...txParams, gasLimit: txParams.gas };
```
Comment (kimpers) [inline subproviders/src/subproviders/private_key_wallet.ts:130]: SUPER NIT:
```suggestion
    const normalizedTxParams = { ...txParams, gasLimit: txParams.gas };
```

**Review by kimpers — APPROVED** [review]
Nice! 🙏

File: subproviders/src/subproviders/private_key_wallet.ts (hunk 7 above)
Comment (dorothy-zbornak) [reply subproviders/src/subproviders/private_key_wallet.ts:130]: thnak yuo

**Pull request merged**